qkeras/quantizers.py

```diff
@@ -439,9 +440,537 @@ def trainable_variables(self):
+      quantization. Special shifting operations are used to achieve this.
+        
+    Example usage:
+
+    # 8-bit quantization with 3 integer bits
+    >>> q = quantized_linear(8, 3)
+    >>> x = tf.constant([0.0, 0.5, 1.0, 1.5, 2.0])
+    >>> q(x).numpy()
+    array([0., 0., 1., 2., 2.], dtype=float32)
+
+    # 2-bit quantization with "auto" and tensor alphas
+    >>> q_auto = quantized_linear(2, alpha="auto")
+    >>> x = tf.constant([0.0, 0.5, 1.0, 1.5, 2.0])
+    >>> q_auto(x).numpy()
+    array([0., 0., 0., 2., 2.], dtype=float32)
+    >>> q_auto.scale.numpy()
+    array([4.], dtype=float32)
+    >>> q_auto.quantization_scale.numpy()
+    array([2.], dtype=float32)
+    >>> q_fixed = quantized_linear(2, alpha=q_auto.scale)
+    >>> q_fixed(x)
+    array([0., 0., 0., 2., 2.], dtype=float32)
+
+    Args:
+      bits (int): Number of bits to represent the number. Defaults to 8.
+      integer (int): Number of bits to the left of the decimal point, used for
+        data_type_scale. Defaults to 0.
+      symmetric (bool): If true, we will have the same number of values 
+        for positive and negative numbers. Defaults to True.
+      alpha (str, Tensor, None): Instructions for determining the quantization
+        scale. Defaults to None.
+        - If None: the quantization scale is the data type scale, determined 
+          by `integer`, `bits`, and `keep_negative`. 
+        - If "auto", the quantization scale is calculated as the minimum 
+          floating point scale per-channel that does not clip the max of x. 
+        - If "auto_po2", the quantization scale is chosen as the
+          power of two per-channel that minimizes squared error between the
+          quantized x and the original x.
+        - If Tensor: The quantization scale is the Tensor passed in
+          multiplied by the data type scale.
+      keep_negative (bool): If false, we clip negative numbers. Defaults to 
+        True.
+      use_stochastic_rounding (bool): If true, we perform stochastic rounding
+        (https://arxiv.org/pdf/1502.02551.pdf).
+      scale_axis (int, None): Which axis to calculate scale from. If None, we
+        perform per-channel scaling based off of the image data format. Note 
+        that each entry of a rank-1 tensor is considered its own channel by
+        default. See `_get_scaling_axis` for more details. Defaults to None.
+      qnoise_factor (float): A scalar from 0 to 1 that represents the level of
+        quantization noise to add. This controls the amount of the
+        quantization noise to add to the outputs by changing the weighted
+        sum of (1 - qnoise_factor) * unquantized_x + qnoise_factor *
+        quantized_x. Defaults to 1.0, which means that the result is fully
+        quantized.
+      use_variables (bool): If true, we use tf.Variables to store certain 
+        parameters. See the BaseQuantizer implementation for more details.
+        Defaults to False.
```

Comment:
Why default to false?

Reply:
1. This is the behavior of quantized_bits
2. It requires the use of special update methods for the qnoise_factor if set to True, so the user should be conscious of making this decision